R/parse-variables-sheet.R

```diff
@@ -0,0 +1,159 @@
+#' Parse and validate a variables sheet
+#'
+#' This function validates that derived variables in a variables sheet do not
+#' use database columns directly. Derived variables must only use non-derived
+#' or other derived variables.
+#'
+#' @param variables_sheet A data frame containing the variables sheet with
+#'   required columns `variable` and `variableStart`
+#'
+#' @return If validation succeeds, returns the variables_sheet as an S3 class
+#'   object with class "variables_sheet". If validation fails, returns a named
+#'   list with `success = FALSE` and an `errors` list containing error details.
+#'   Each error is a named list with fields: `type` (error identifier), `row`
+#'   (row index), and `message` (error description).
+#'
+#' @export
+parse_variables_sheet <- function(variables_sheet) {
+  # Input validation
+  .validate_variables_sheet_input(variables_sheet)
+
+  # Core validation logic
+  errors <- .validate_derived_variables(variables_sheet)
+
+  # Return success or failure
+  if (length(errors) > 0) {
+    return(list(
+      success = FALSE,
+      errors = errors
+    ))
+  } else {
+    class(variables_sheet) <- c("variables_sheet", class(variables_sheet))
+    return(variables_sheet)
+  }
+}
+
+#' Validate input parameter for parse_variables_sheet
+#'
+#' @param variables_sheet The input to validate
+#'
+#' @return TRUE if validation passes. Throws errors otherwise.
+#' @keywords internal
+.validate_variables_sheet_input <- function(variables_sheet) {
+  # Check that it's a data frame
+  if (!checkmate::test_data_frame(variables_sheet)) {
+    cli::cli_abort(c(
+      "{.arg variables_sheet} must be a data.frame",
+      "x" = "{.arg variables_sheet} is a {typeof(variables_sheet)}"
+    ))
+  }
+
+  # Check that required columns exist
+  required_cols <- c(
+    pkg.env$columns.variable,
+    pkg.env$columns.variableStart
+  )
+
+  if (!checkmate::test_names(colnames(variables_sheet), must.include = required_cols)) {
+    missing_cols <- required_cols[!required_cols %in% names(variables_sheet)]
+    cli::cli_abort(c(
+      "{.arg variables_sheet} is missing required columns",
+      "x" = "{missing_cols} are missing"
+    ))
+  }
+
+  return(TRUE)
+}
+
+#' Validate that derived variables do not use database columns directly
+#'
+#' @param variables_sheet A data frame containing the variables sheet
+#'
+#' @return A list of errors (empty if no errors found)
+#' @keywords internal
+.validate_derived_variables <- function(variables_sheet) {
+  errors <- list()
+
+  # If the data frame is empty, return no errors
+  if (nrow(variables_sheet) == 0) {
+    return(errors)
+  }
+
+  # Iterate through each row
+  for (i in seq_len(nrow(variables_sheet))) {
```

Comment:
Don't use mutation on the errors variable. Instead use a function like purrr::map or purrr::reduce